lib/active_storage/service/cloudinary_service.rb

```diff
@@ -0,0 +1,156 @@
+    def upload(key, io, checksum: nil)
+      instrument :upload, key: key, checksum: checksum do
+        Cloudinary::Uploader.upload(io, public_id: key)
+      end
+    end
+
+    # Return the content of the file at the +key+.
+    def download(key)
+      tmp_file = open(url_for_public_id(key))
+      if block_given?
+        instrument :streaming_download, key: key do
+          File.open(tmp_file, 'rb') do |file|
+            while (data = file.read(64.kilobytes))
+              yield data
+            end
+          end
+        end
+      else
+        instrument :download, key: key do
+          File.binread tmp_file
+        end
+      end
+    end
+
+    # Delete the file at the +key+.
+    def delete(key)
+      instrument :delete, key: key do
+        delete_resource_with_public_id(key)
+      end
+    end
+
+    # Delete files at keys starting with the +prefix+.
+    def delete_prefixed(prefix)
+      instrument :delete_prefixed, prefix: prefix do
+        find_resources_with_public_id_prefix(prefix).each do |resource|
+          delete_resource_with_public_id(resource['public_id'])
+        end
+      end
+    end
+
+    # Return +true+ if a file exists at the +key+.
+    def exist?(key)
+      instrument :exist?, key: key do
+        resource_exists_with_public_id?(key)
+      end
+    end
+
+    # Returns a signed, temporary URL for the file at the +key+. The URL will be valid for the amount
+    # of seconds specified in +expires_in+. You most also provide the +disposition+ (+:inline+ or +:attachment+),
+    # +filename+, and +content_type+ that you wish the file to be served with on request.
+    def url(key, expires_in:, disposition:, filename:, content_type:)
+      instrument :url, key: key do
+        options = {
+          expires_in: expires_in,
+          content_type: content_type,
+          disposition: disposition,
+          filename: filename
+        }
+        signed_download_url_for_public_id(key, options)
+      end
+    end
+
+    # Returns a signed, temporary URL that a direct upload file can be PUT to on the +key+.
+    # The URL will be valid for the amount of seconds specified in +expires_in+.
+    # You must also provide the +content_type+, +content_length+, and +checksum+ of the file
+    # that will be uploaded. All these attributes will be validated by the service upon upload.
+    def url_for_direct_upload(key, expires_in:, content_type:, content_length:, checksum:)
+      instrument :url_for_direct_upload, key: key do
+        options = {
+          expires_in: expires_in,
+          content_type: content_type,
+          content_length: content_length,
+          checksum: checksum
+        }
+        direct_upload_url_for_public_id(key, options)
+      end
+    end
+
+    # Returns a Hash of headers for +url_for_direct_upload+ requests.
+    def headers_for_direct_upload(key, filename:, content_type:, content_length:, checksum:)
+      { 'Content-Type' => content_type, 'X-Unique-Upload-Id' => key }
```

Comment:
since this is only using 2 arguments, would it be better to just provide that. Also, do you think this should be a public API?

Reply:
Yea, active_storage calls into when building out attributes for client side direct upload. I did a poor job with the PR description, should've included an overview of the active_storage interface and workings 😅 

Update: Add link to the interface 😄